src/mongo_types.ts

```diff
@@ -258,7 +258,26 @@ export type OnlyFieldsOfType<TSchema, FieldType = any, AssignableType = FieldTyp
 >;
 
 /** @public */
-export type MatchKeysAndValues<TSchema> = Readonly<Partial<TSchema>> & Record<string, any>;
+export type MatchKeysAndValues<TSchema> = Readonly<{
+  [Property in Join<NestedPaths<TSchema>, '.'>]?: PropertyType<TSchema, Property>;
+}> &
```

Comment:
The `ReadOnly<...>` could be extracted, eg:

```
type MatchKeysAndValuesNoPositional<TSchema> =  Readonly<{
  [Property in Join<NestedPaths<TSchema>, '.'>]?: PropertyType<TSchema, Property>;
}>

type MatchKeysAndValues<TSchema> = MatchKeysAndValuesNoPositional<TSchema> & Partial<
    Record<
      `${KeysOfAType<MatchKeysAndValuesNoPositional<TSchema>, any[]>}.$${`[${string}]` | ''}${
        // Subsequent nested path - could be narrowed
        `.${string}` | ''
      }`,
      // Value for updated field - could be narrowed
      any
    >
  >;
```

But that would make an additional exported type